drivers/fpga/fpga_altera.c

```diff
@@ -0,0 +1,463 @@
+ * @param[in] function_identifier Function identifier for each command type
+ * @param[in] cmd_request
+ * @param[in] private_data
+ *
+ * @return 0 on success or negative value on fail
+ */
+static int32_t smc_send(uint32_t cmd_type, uint64_t function_identifier, uint32_t *cmd_request,
+			struct private_data_t *private_data)
+{
+	int32_t trans_id = 0;
+	uint32_t *cmd_addr = NULL;
+	uint32_t *resp_addr = NULL;
+	struct sip_svc_request request;
+
+	if (!mailbox_smc_dev) {
+		LOG_ERR("Mailbox client is not registered");
+		return -ENODEV;
+	}
+
+	if (cmd_type == SIP_SVC_PROTO_CMD_ASYNC) {
+		cmd_addr = (uint32_t *)k_malloc(FPGA_MB_CMD_ADDR_MEM_SIZE);
+		if (!cmd_addr) {
+			LOG_ERR("Failed to allocate command memory");
+			return -ENOMEM;
+		}
+		cmd_addr[MBOX_CMD_HEADER_INDEX] =
+			MBOX_REQUEST_HEADER(cmd_request[SMC_REQUEST_A2_INDEX], 0, 0);
+		resp_addr = (uint32_t *)k_malloc(FPGA_MB_RESPONSE_MEM_SIZE);
+		if (!resp_addr) {
+			k_free(cmd_addr);
+			return -ENOMEM;
+		}
+
+		request.a2 = (uint64_t)cmd_addr;
+		request.a3 = sizeof(uint32_t);
+		request.resp_data_addr = (uint64_t)resp_addr;
+		request.resp_data_size = FPGA_MB_RESPONSE_MEM_SIZE;
+
+#if defined(CONFIG_LOG)
+		for (int32_t mbox_idx = 0; mbox_idx < request.a3 / 4; mbox_idx++) {
+			LOG_DBG("\t [%d ] %08x", mbox_idx, cmd_addr[mbox_idx]);
+		}
+#endif
+
+	} else {
+		request.a2 = cmd_request[SMC_REQUEST_A2_INDEX];
+		request.a3 = cmd_request[SMC_REQUEST_A3_INDEX];
+		request.resp_data_addr = 0;
+		request.resp_data_size = 0;
+	}
+
+	/* Fill SiP SVC request buffer */
+	request.header = SIP_SVC_PROTO_HEADER(cmd_type, 0);
+	request.a0 = function_identifier;
+	request.a1 = 0;
+	request.a4 = 0;
+	request.a5 = 0;
+	request.a6 = 0;
+	request.a7 = 0;
+	request.priv_data = (void *)private_data;
+
+	/* Send SiP SVC request */
+	trans_id = sip_svc_send(mailbox_smc_dev, mailbox_client_token, &request, smc_callback);
+
+	if (trans_id == SIP_SVC_ID_INVALID) {
+		LOG_ERR("SiP SVC send request fail");
+		return -EBUSY;
+	}
+
+	return 0;
+}
+
+/* Validate the Reconfig status response */
+static int32_t fpga_reconfig_status_validate(struct fpga_config_status *reconfig_status_resp)
+{
+	uint32_t ret = 0;
+
+	/* Check for any error */
+	ret = reconfig_status_resp->state;
+	if (ret == MBOX_CFGSTAT_VAB_BS_PREAUTH) {
+		return MBOX_CONFIG_STATUS_STATE_CONFIG;
+	}
+
+	if (ret && ret != MBOX_CONFIG_STATUS_STATE_CONFIG)
+		return ret;
+
+	/* Make sure nStatus is not 0 */
+	ret = reconfig_status_resp->pin_status.pin_status;
+	if (!(ret & RECONFIG_PIN_STATUS_NSTATUS))
+		return MBOX_CFGSTAT_STATE_ERROR_HARDWARE;
+
+	ret = reconfig_status_resp->soft_function_status;
+	if ((ret & RECONFIG_SOFTFUNC_STATUS_CONF_DONE) &&
+	    (ret & RECONFIG_SOFTFUNC_STATUS_INIT_DONE) && !reconfig_status_resp->state)
+		return 0; /* Configuration success */
+
+	return MBOX_CONFIG_STATUS_STATE_CONFIG;
+}
+
+/* Will send the SMC command to check the status of the FPGA */
+static int32_t is_fpga_config_not_ready(void)
```

Comment:
It might be better to rename this so that it provides non-inverted logical operation (adjust return value as necessary).

Reply:
done